File: content/blog/longest_valid_parentheses/go/solution_stack.go
package longest_valid_parentheses

func longestValidParenthesesStack(s string) int {
	stack := make([]int, 0, len(s))
	stack = append(stack, -1)
	res := 0

	for i := 0; i < len(s); i++ {
		if s[i] == '(' {
			stack = append(stack, i)
			continue
		}

		stack = stack[:len(stack)-1]
		if len(stack) == 0 {
			stack = append(stack, i)
			continue
		}

		res = max(res, i-stack[len(stack)-1])
	}

	return res
}

File: content/blog/longest_valid_parentheses/ts/solution_stack.ts
export function longestValidParenthesesStack(s: string): number {
    let stack: number[] = [-1];
    let res = 0;

    for (let i = 0; i < s.length; i++) {
        if (s[i] === '(') {
            stack.push(i);
            continue;
        }

        stack.pop();
        if (stack.length === 0) {
            stack.push(i);
            continue;
        }

        res = Math.max(res, i - stack[stack.length - 1]);
    }

    return res;
}
